File: C++/euler-204-generalised-hamming-numbers.cpp/euler-204-generalised-hamming-numbers.cpp
#include <iostream>   // Input and output stream handling (for std::cout and std::endl)
#include <cmath>      // Math functions like sqrt used in isPrimeNumber function
#include <vector>     // Dynamic array for storing prime factors in the HammingNumbers class
#include <cstdint>    // Standard integer types for fixed-size integers, used for int64_t as integers have larger values in hammingNumbers

class HammingNumbers {
private:
    std::vector<int> factors;  // Stores prime factors up to the given limit
    int limit;  // The limit up to which Hamming numbers are calculated

public:
    HammingNumbers() {}

    // Method to initiate the calculation of Hamming numbers
    void run(int primetype, int limit) {
        this->limit = limit;

        // Find prime factors up to the given limit
        for (int num = 1; num <= primetype; num++) {
            if (isPrimeNumber(num)) {
                factors.push_back(num);
            }
        }

        // Initialize the totalCount and calculate combinations starting from the first factor
        int64_t finalCount = hammingNumbers(0, 1, 0);

        // Print the final count of combinations that satisfied the condition
        std::cout << finalCount << std::endl;
    }

private:
    // Function to check if a number is prime
    bool isPrimeNumber(int number) {
        if (number < 2) {
            return false;
        }
        int squareRoot = static_cast<int>(std::sqrt(static_cast<double>(number)));
        for (int i = 2; i <= squareRoot; i++) {
            if (number % i == 0) {
                return false;
            }
        }
        return true;
    }

    // Recursive function to calculate Hamming numbers
    int64_t hammingNumbers(int index, int64_t product, int64_t total) {
        // If the current product is within the limit, increment the total count
        if (product <= limit) {
            total++; // Increment the total count for valid combinations

            // Recursively calculate combinations for the remaining factors
            for (int j = index; j < factors.size(); j++) {
                total = hammingNumbers(j, product * factors[j], total);
            }
        }

        return total;
    }
};

int main(int argc, char *argv[]) {
    // Check if there are two values inputted
    if (argc != 3) {
        std::cout << "Usage: ./euler-204-generalised-hamming-numbers <type> <limit>" << std::endl;
        return 0;
    }

    char* endptr1;
    char* endptr2;

    // Convert command line arguments to integers
    int primetype = std::strtol(argv[1], &endptr1, 10);
    int limit = std::strtol(argv[2], &endptr2, 10);

    // Check if the conversion was successful
    if (*endptr1 != '\0' || *endptr2 != '\0') {
        std::cout << "Error: Both <type> and <limit> must be integers." << std::endl;
        return 0;
    }

    // Create an instance of HammingNumbers
    HammingNumbers numbers;

    // Run the calculation
    numbers.run(primetype, limit);

    return 0;
}

File: Go/euler-204-generalised-hamming-numbers.go/euler-204-generalised-hamming-numbers.go
package main

import (
	"fmt"     // Fmt is used to print a formatted string to the console
	"math"    // Math provides mathematical functions. In this program, it is used for logarithmic calculations.
	"os"      // Os is used to access the command line arguments and to terminate the program with a specified exit code, this is for error handling
	"strconv" // Strconv is used to convert a string to an integer from the command line
)

// factors represent the prime factors to be used in the combinations
var factors []int

// limit is the upper limit for the product in the combinations
var limit int

func main() {

	// Check if there are two values inputted
	if len(os.Args) != 3 {
		fmt.Println("Usage: go run euler-204-generalised-hamming-numbers.go <type> <limit>")
		return
	}

	// Prime Type
	primetype, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Printf("Invalid input: Please enter a valid type.\nError message: %v\n", err)
		return
	}

	// Limit
	limit, err = strconv.Atoi(os.Args[2])
	if err != nil {
		fmt.Printf("Invalid input: Please enter a valid limit.\nError message: %v\n", err)
		return
	}

	// Find prime factors up to the given limit
	for num := 1; num <= primetype; num++ {
		result := isPrimeNumber(num)
		if result {
			factors = append(factors, num)
		}
	}

	// Initialize the totalCount and calculate combinations starting from the first factor
	final := generalisedHammingNumbers(0, 1, 0)

	// Print the final count of combinations that satisfied the condition
	fmt.Println(final)
}

// isPrimeNumber checks if a given number is a prime number
func isPrimeNumber(number int) bool {
	if number < 2 {
		return false
	}
	squareRoot := int(math.Sqrt(float64(number)))
	for i := 2; i <= squareRoot; i++ {
		if number%i == 0 {
			return false
		}
	}
	return true
}

// generalisedHammingNumbers recursively computes the combinations of prime factors
// to find the count of generalised Hamming numbers within a specified limit
// Parameters:
//
//	index: current index in the prime factor array
//	product: current product of the prime factors in the combination
//	total: total count of combinations if satisfying the condition
func generalisedHammingNumbers(index, product, total int) int {
	// If the current product is within the limit, increment the total count
	if product <= limit {
		total++ // Increment the total count for valid combinations

		// Recursively calculate combinations for the remaining factors
		for j := index; j < len(factors); j++ {
			total = generalisedHammingNumbers(j, product*factors[j], total)
		}
	}

	return total
}
